src/Compilers/CSharp/Portable/Compilation/MethodBodySemanticModel.cs

```diff
@@ -19,32 +19,45 @@ private MethodBodySemanticModel(CSharpCompilation compilation, Symbol owner, Bin
         }
 
         /// <summary>
-        /// Creates a SemanticModel that creates and owns the ExecutableCodeBinder for the method of which it is a model.
+        /// Creates a SemanticModel for the method of which it is a model.
         /// </summary>
-        internal static MethodBodySemanticModel Create(CSharpCompilation compilation, MethodSymbol owner, Binder rootBinder, CSharpSyntaxNode syntax)
+        internal static MethodBodySemanticModel Create(CSharpCompilation compilation, MethodSymbol owner, ExecutableCodeBinder executableCodeBinder, 
+                                                       CSharpSyntaxNode syntax, BoundNode boundNode = null)
         {
-            var executableCodeBinder = new ExecutableCodeBinder(syntax, owner, rootBinder);
-            return new MethodBodySemanticModel(compilation, owner, executableCodeBinder, syntax);
-        }
+            var result =  new MethodBodySemanticModel(compilation, owner, executableCodeBinder, syntax);
 
-        internal override BoundNode Bind(Binder binder, CSharpSyntaxNode node, DiagnosticBag diagnostics)
-        {
-            if (node.Kind() == SyntaxKind.ArrowExpressionClause)
+            if (boundNode != null)
             {
-                return binder.BindExpressionBodyAsBlock((ArrowExpressionClauseSyntax)node, diagnostics);
+                result.UnguardedAddBoundTreeForStandaloneSyntax(syntax, boundNode);
             }
-            return base.Bind(binder, node, diagnostics);
+
+            return result;
         }
 
-        internal override BoundNode GetBoundRoot()
+        internal override BoundNode Bind(Binder binder, CSharpSyntaxNode node, DiagnosticBag diagnostics)
         {
-            CSharpSyntaxNode root = this.Root;
-            if (root.Kind() == SyntaxKind.ArrowExpressionClause)
+            switch (node.Kind())
             {
-                root = ((ArrowExpressionClauseSyntax)root).Expression;
-                return GetUpperBoundNode(GetBindableSyntaxNode(root));
+                case SyntaxKind.ArrowExpressionClause:
+                    return binder.BindExpressionBodyAsBlock((ArrowExpressionClauseSyntax)node, diagnostics);
```

Comment:
It would be good to constrain this somehow to ensure it the the arrow-expression-clause at the top level. We will be using the ArrowExpressionClause elsewhere in the syntax (e.g. the switch expression).

Reply:
> It would be good to constrain this somehow to ensure it the the arrow-expression-clause at the top level. We will be using the ArrowExpressionClause elsewhere in the syntax (e.g. the switch expression).

Those won't be bindable on their own, I assume, therefore we should not be attempting binding them as stand-alone nodes in SemanticModel. I will try to add an assert, but you will have to follow up on SemanticModel behavior for the switch expression as part of that feature. As it stands right now, I believe the handling of ArrowExpressionClause is correct. 

---
In reply to: [170038212](https://github.com/dotnet/roslyn/pull/24918#discussion_r170038212) [](ancestors = 170038212)

Reply:
>  I will try to add an assert

Actually, I don't believe any assert is appropriate at the moment as we might be binding a body of a local function in an error recovery mode, i.e when we have a block body and an expression body at the same time.

---
In reply to: [170084715](https://github.com/dotnet/roslyn/pull/24918#discussion_r170084715) [](ancestors = 170084715,170038212)